src/mm/page_manager.rs

```diff
@@ -0,0 +1,214 @@
+    Reserved,
+    Kernel,
+    Used,
+    Unused,
+}
+
+/// Holds data about each physical page in the system.
+#[derive(Debug)]
+pub struct PhysicalPage {
+    kind: PageKind,
+    base: usize,
+
+    /// Page is last of a contiguous allocation of pages.
+    last: bool, 
+}
+
+impl PhysicalPage {
+    fn is_unused(&self) -> bool {
+        self.kind == PageKind::Unused
+    }
+
+    fn is_used(&self) -> bool {
+        self.kind != PageKind::Unused
+    }
+
+    fn set_used(&mut self) {
+        self.kind = PageKind::Used;
+    }
+
+    fn _is_last(&self) -> bool {
+        self.last
+    }
+
+    fn set_last(&mut self) {
+        self.last = true;
+    }
+
+    pub fn base(&self) -> usize {
+        self.base
+    }
+}
+
+
+pub struct PageManager<'a> {
+    metadata: &'a mut [PhysicalPage],
+    page_size: usize,
+}
+
+impl<'a> PageManager<'a> {
+
+    fn count_pages(regions: impl Iterator<Item = fdt::standard_nodes::MemoryRegion>, page_size: usize) -> usize {
+        regions
+            .map(|region| {
+                (region.starting_address, unsafe {
+                    region.starting_address.add(region.size.unwrap_or(0))
+                })
+            })
+            .map(|(start, end)| (start as usize, end as usize))
+            .flat_map(|(start, end)| (start..end).step_by(page_size))
+            .count()
+    }
+
+    fn align_up(addr: usize, alignment: usize) -> usize {
+        ((addr) + alignment - 1) & !(alignment - 1)
+    }
+
+    fn phys_addr_to_physical_page(phys_addr: usize, device_tree: &fdt::Fdt, page_size: usize) -> PhysicalPage {
+        let kind = if mm::is_kernel_page(phys_addr) {
+            PageKind::Kernel
+        } else if mm::is_reserved_page(phys_addr, device_tree, page_size) {
+            PageKind::Reserved
+        } else {
+            PageKind::Unused
+        };
+
+        PhysicalPage {
+            kind,
+            base: phys_addr,
+            last: false,
+        }
+    }
+
+    /// Look for `pages_needed` contiguous unused pages, beware of pages that are in use by the
+    /// kernel.
+    fn find_contiguous_unused_pages(device_tree: &fdt::Fdt,
+    pages_needed: usize, page_size: usize) -> Option<usize> {
+        let memory = device_tree.memory();
+
+        let physical_pages = memory.regions()
+            .map(|region| {
+                (region.starting_address, unsafe {
+                    region.starting_address.add(region.size.unwrap_or(0))
+                })
+            })
+            .map(|(start, end)| (start as usize, end as usize))
+            .flat_map(|(start, end)| (start..end).step_by(page_size))
+            .map(|base| Self::phys_addr_to_physical_page(base, device_tree, page_size));
+
+        let mut first_page_addr: usize = 0;
+        let mut consecutive_pages: usize = 0;
+
+        for page in physical_pages {
+            if consecutive_pages == 0 {
+                first_page_addr = page.base;
+            }
+
+            if !page.is_unused() {
```

Comment:
```suggestion
            if page.is_used() {
```
I find this more readable